### DISCUSSION TITLE
Kotlin 2.0 & K2, what does this mean for Detekt?

### DISCUSSION THREAD

**Discussion by TWiStErRob** [detekt/detekt#5775] [Ideas]
Hi there, the news is out, K2 is coming soon (this or next year):
https://blog.jetbrains.com/kotlin/2023/02/k2-kotlin-2-0/

What does this mean for Detekt?
Is there a need for rewrite (PSI changes)?
Will we just get "free" performance?
Will it break the [compiler plugin](https://github.com/detekt/detekt/pull/5492)?
Will there be an actual [Detekt 2.0](https://github.com/detekt/detekt/milestone/42)? 😊

**Comment by schalkms** [comment]
The K2 compiler gets a new frontend called frontend intermediate representation. As detekt relies on the PSI, this will most likely impact further development after K2 has been stable.

https://twitter.com/kotlin/status/1453742513379020800

**Reply by TWiStErRob** [reply]
Related nice summary in form a blogpost: https://akuleshov7.com/2021-06-24-kotlin-representation.html

**Comment by picklebento** [comment]
We could try with K2 earlier following the documentation to provide early feedback https://kotlinlang.org/docs/whatsnew-eap.html?_gl=1*1pf7wr0*_ga*MjA0MzU5OTIxMy4xNjc3NDQ4MzE3*_ga_9J976DJZ68*MTY3NzQ0ODMxNy4xLjEuMTY3NzQ0ODM4NS41NS4wLjA.&_ga=2.180902631.1599569198.1677448318-2043599213.1677448317#new-kotlin-k2-compiler-updates

**Comment by cortinico** [comment]
> Will there be an actual [Detekt 2.0](https://github.com/detekt/detekt/milestone/42)? 😊

Sneak peak: yes :D We're working on it <3

And yes, I believe that K2 would mean a lot of rules will have to be re-written or adapted.
Ideally we could make a community effort out of it.
